src/lem/circuit.rs

```diff
@@ -56,9 +58,157 @@ use super::{
+        // `count` is the number of `Frames` to include per `MultiFrame`.
+        let total_frames = frames.len();
+        let n = total_frames / count + usize::from(total_frames % count != 0);
+        let mut multi_frames = Vec::with_capacity(n);
+
+        for chunk in frames.chunks(count) {
+            let mut inner_frames = Vec::with_capacity(count);
+
+            for frame in chunk {
+                inner_frames.push(frame.clone());
+            }
+
+            let last_frame = chunk.last().expect("chunk must not be empty");
+            let last_circuit_frame = inner_frames
+                .last()
+                .expect("chunk must not be empty")
+                .clone();
+
+            // Fill out the MultiFrame, if needed, and capture output of the final actual frame.
+            for _ in chunk.len()..count {
+                inner_frames.push(last_circuit_frame.clone());
+            }
+
+            let output = last_frame.output.clone();
+            let input = chunk[0].input.clone();
+            debug_assert!(!inner_frames.is_empty());
+
+            let mf = MultiFrame {
+                func,
+                store: Some(store),
+                input: Some(input),
+                output: Some(output),
+                frames: Some(inner_frames),
+                cached_witness: None,
+                count,
+            };
+
+            multi_frames.push(mf);
+        }
+
+        multi_frames
+    }
+
+    /// Make a dummy `MultiFrame`, duplicating `self`'s final `CircuitFrame`.
+    pub(crate) fn make_dummy(
+        func: &'a Func,
+        count: usize,
+        circuit_frame: Option<Frame<F>>,
+        store: &'a Store<F>,
+    ) -> Self {
+        let (frames, input, output) = if let Some(circuit_frame) = circuit_frame {
+            (
+                Some(vec![circuit_frame.clone(); count]),
+                Some(circuit_frame.input),
+                Some(circuit_frame.output),
+            )
+        } else {
+            (None, None, None)
+        };
+        Self {
+            func,
+            store: Some(store),
+            input,
+            output,
+            frames,
+            cached_witness: None,
+            count,
+        }
+    }
+
+    pub fn synthesize_frames<CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        store: &Store<F>,
+        input: &[AllocatedPtr<F>],
+        frames: &[Frame<F>],
+        blank: bool,
+    ) -> Vec<AllocatedPtr<F>> {
+        let global_allocator = &mut GlobalAllocator::default();
+        let (_, output) = frames
+            .iter()
+            .fold((0, input.to_vec()), |(i, input), frame| {
+                if !blank {
+                    for (alloc_ptr, input) in input.iter().zip(&frame.input) {
+                        let input_zptr = store.hash_ptr(input).expect("Hash did not succeed");
+                        assert_eq!(
+                            alloc_ptr.tag().get_value().expect("Assignment missing"),
```

Comment:
`synthesize_frames` should return `Result<Vec<AllocatedPtr<F>>, SynthesisError>`, and this should return a `SynthesisError::AssignmentMissing` rather than panic.